yktracec/src/jitmodbuilder.cc

```diff
@@ -1109,14 +1115,27 @@ class JITModBuilder {
           continue;
         } else if (LastInst && isa<ReturnInst>(LastInst)) {
           LastInst = nullptr;
-          assert(CallStack.back()->getParent() == BB);
-          LastBB = BB;
+          if (!IsSWTrace) {
+            assert(CallStack.back()->getParent() == BB);
+          }
+          LastBB = CallStack.back()->getParent();
```

Comment:
This is the main remaining line in the PR that I don't understand: any clues?

Reply:
In the hwt we see blocks containing calls twice, once before the call happens, and then when we return from the call. This is due to how PT sees things. So this assert basically checks, that after returning from a call, we end up in the same block as before the call (this block is then skipped immediately as we've already processed it before the call). In the software tracer we don't see blocks twice, so when we return from a call, the next block we see is the block following the block containing the call. So this assert is incorrect in that case. It's just a small hwt sanity check, but it can probably be removed.

Reply:
Got it. I think we should remove the `assert` from jitmodbuilder. I wonder if we can work out how to either a) `assert` the same thing in hwt b) get hwt to merge the blocks such that jitmodbuilder sees swt and hwt as doing the same thing? That is probably somewhat hard, but worth thinking about, as it will keep the code simpler if jitmodbuilder (or jitc_yk) can be ignorant of the tracing backend.